VirtualFilesystemDirect.php

```diff
@@ -255,6 +255,84 @@ public function delete( $fileOrDir, $recursive = false, $type = false ) {
 		return @unlink( $this->getUrl( $fileOrDir ) );
 	}
 
+	/**
+	 * Gets details for files in a directory or a specific file.
+	 *
+	 * Copied from WP_Filesystem_Direct.
+	 *
+	 * @param string $path           Path to directory or file.
+	 * @param bool   $include_hidden Optional. Whether to include details of hidden ("." prefixed) files.
+	 *                               Default true.
+	 * @param bool   $recursive      Optional. Whether to recursively include file details in nested directories.
+	 *                               Default false.
+	 *
+	 * @return array|false Array of files. False if unable to list directory contents.
+	 */
+	public function dirlist( $path, $include_hidden = true, $recursive = false ) {
+		$path = $this->getUrl( $path );
+
+		if ( $this->is_file( $path ) ) {
+			$limit_file = basename( $path );
+			$path       = dirname( $path );
+		} else {
+			$limit_file = false;
+		}
+
+		if ( ! $this->is_dir( $path ) || ! $this->is_readable( $path ) ) {
+			return false;
+		}
+
+		$dir = dir( $path );
+		if ( ! $dir ) {
+			return false;
+		}
+
+		$listing = [];
+
+		while ( false !== ( $entry = $dir->read() ) ) {
+			$struc         = [
+				'name' => $entry,
+			];
+
+			if ( '.' === $struc['name'] || '..' === $struc['name'] ) {
+				continue;
+			}
+
+			if ( ! $include_hidden && '.' === $struc['name'][0] ) {
+				continue;
+			}
+
+			if ( $limit_file && $struc['name'] !== $limit_file ) {
+				continue;
+			}
+
+			$path                 = rtrim( $path, '\//' );
+			$pathentry            = "{$path}/{$entry}";
+			$struc['perms']       = $this->gethchmod( $pathentry );
+			$struc['permsn']      = $this->getnumchmodfromh( $struc['perms'] );
+			$struc['number']      = false;
+			$struc['owner']       = $this->owner( $pathentry );
+			$struc['group']       = $this->group( $pathentry );
+			$struc['size']        = $this->size( $pathentry );
+			$struc['lastmodunix'] = $this->mtime( $pathentry );
+			$struc['lastmod']     = gmdate( 'M j', $struc['lastmodunix'] );
+			$struc['time']        = gmdate( 'h:i:s', $struc['lastmodunix'] );
+			$struc['type']        = $this->is_dir( $pathentry ) ? 'd' : 'f';
+
+			if ( 'd' === $struc['type'] ) {
+				$struc['files'] = $recursive
+					? $this->dirlist( $path . '/' . $struc['name'], $include_hidden, $recursive )
+					: [];
+			}
+
+			$listing[ $struc['name'] ] = $struc;
```

Comment:
@hellofromtonya I wonder if we should have this:
```php
$listing[ $struc['name'] ] = $struc;

if ( $limit_file ) {
	break;
}
```

The reason: if I understand this correctly, passing a file path as first argument for this method (not a directory path) means we want info only about this file. In this case, there is no need to continue to loop through the remaining files once we have the one we want.
Does it make sense or do I have it wrong?
And in case I'm right, should we even roll away from the original WP’s function?

Reply:
That is true. That would be a good optimization change for Core!

I tried to keep the functionality the same as [`VirtualFIlesystemDirect:: dirlist()`](https://github.com/WordPress/WordPress/blob/641c632b0c9fde4e094b217f50749984ca43a2fa/wp-admin/includes/class-wp-filesystem-direct.php#L560).

Reply:
Ok, let’s stick to the core.